File: engines/go/inserter_pgsql.go
package main

import (
	"bytes"
	"context"
	"database/sql"
	"encoding/csv"
	"fmt"
	"strings"

	"github.com/jackc/pgx/v5"
	_ "github.com/jackc/pgx/v5/stdlib"
)

type PgSQLInserter struct {
	db     *sql.DB
	params ConnParams
}

func NewPgSQLInserter(params ConnParams) (*PgSQLInserter, error) {
	dsn := fmt.Sprintf(
		"host=%s port=%d user=%s password=%s dbname=%s sslmode=disable",
		params.Host, params.Port, params.User, params.Password, params.Database,
	)
	db, err := sql.Open("pgx", dsn)
	if err != nil {
		return nil, fmt.Errorf("sql.Open: %w", err)
	}
	if err := db.Ping(); err != nil {
		return nil, fmt.Errorf("ping: %w", err)
	}
	return &PgSQLInserter{db: db, params: params}, nil
}

func (ins *PgSQLInserter) Close() {
	ins.db.Close()
}

func (ins *PgSQLInserter) quote(name string) string {
	clean := cleanStr(name)
	return `"` + strings.ReplaceAll(clean, `"`, `""`) + `"`
}

func (ins *PgSQLInserter) placeholder(i int) string {
	return fmt.Sprintf("$%d", i+1)
}

func (ins *PgSQLInserter) DefaultInsert(csvFile, tableName string) (int, error) {
	data, err := readCSV(csvFile)
	if err != nil {
		return 0, err
	}

	cols := make([]string, len(data.Headers))
	phs  := make([]string, len(data.Headers))
	for i, h := range data.Headers {
		cols[i] = ins.quote(h)
		phs[i]  = ins.placeholder(i)
	}

	query := fmt.Sprintf(
		"INSERT INTO %s (%s) VALUES (%s)",
		ins.quote(tableName),
		strings.Join(cols, ", "),
		strings.Join(phs, ", "),
	)

	tx, err := ins.db.Begin()
	if err != nil {
		return 0, fmt.Errorf("begin tx: %w", err)
	}

	stmt, err := tx.Prepare(query)
	if err != nil {
		tx.Rollback()
		return 0, fmt.Errorf("prepare stmt: %w", err)
	}
	defer stmt.Close()

	for _, row := range data.Rows {
		if _, err := stmt.Exec(rowToArgs(row)...); err != nil {
			tx.Rollback()
			return 0, fmt.Errorf("exec row: %w", err)
		}
	}

	return len(data.Rows), tx.Commit()
}

func (ins *PgSQLInserter) BulkInsert(csvFile, tableName string, batchSize int) (int, error) {
	data, err := readCSV(csvFile)
	if err != nil {
		return 0, err
	}

	cols := make([]string, len(data.Headers))
	for i, h := range data.Headers {
		cols[i] = ins.quote(h)
	}
	colStr := strings.Join(cols, ", ")
	table  := ins.quote(tableName)
	total  := 0

	for i := 0; i < len(data.Rows); i += batchSize {
		end := i + batchSize
		if end > len(data.Rows) {
			end = len(data.Rows)
		}
		batch := data.Rows[i:end]

		valueStrings := make([]string, len(batch))
		valueArgs    := make([]interface{}, 0, len(batch)*len(data.Headers))

		for j, row := range batch {
			phs := make([]string, len(data.Headers))
			for k := range data.Headers {
				phs[k] = ins.placeholder(j*len(data.Headers) + k)
			}
			valueStrings[j] = "(" + strings.Join(phs, ", ") + ")"
			valueArgs = append(valueArgs, rowToArgs(row)...)
		}

		query := fmt.Sprintf(
			"INSERT INTO %s (%s) VALUES %s",
			table, colStr, strings.Join(valueStrings, ", "),
		)

		tx, err := ins.db.Begin()
		if err != nil {
			return total, fmt.Errorf("begin tx: %w", err)
		}
		if _, err := tx.Exec(query, valueArgs...); err != nil {
			tx.Rollback()
			return total, fmt.Errorf("exec batch: %w", err)
		}
		if err := tx.Commit(); err != nil {
			return total, fmt.Errorf("commit: %w", err)
		}
		total += len(batch)
	}

	return total, nil
}

func (ins *PgSQLInserter) FileInsert(csvFile, tableName string) (int, error) {
	data, err := readCSV(csvFile)
	if err != nil {
		return 0, err
	}

	// Формируем CSV в памяти с очищенными заголовками
	var buf bytes.Buffer
	writer := csv.NewWriter(&buf)

	cleanHeaders := make([]string, len(data.Headers))
	for i, h := range data.Headers {
		cleanHeaders[i] = cleanStr(h)
	}
	writer.Write(cleanHeaders)
	writer.WriteAll(data.Rows)
	writer.Flush()

	// Прямое pgx соединение для COPY FROM STDIN
	ctx := context.Background()
	dsn := fmt.Sprintf(
		"host=%s port=%d user=%s password=%s dbname=%s sslmode=disable",
		ins.params.Host, ins.params.Port,
		ins.params.User, ins.params.Password, ins.params.Database,
	)

	conn, err := pgx.Connect(ctx, dsn)
	if err != nil {
		return 0, fmt.Errorf("pgx connect: %w", err)
	}
	defer conn.Close(ctx)

	copySQL := fmt.Sprintf(
		"COPY %s FROM STDIN WITH (FORMAT csv, HEADER true)",
		ins.quote(tableName),
	)

	tag, err := conn.PgConn().CopyFrom(ctx, &buf, copySQL)
	if err != nil {
		return 0, fmt.Errorf("copy from stdin: %w", err)
	}

	return int(tag.RowsAffected()), nil
}

File: engines/rust/src/inserter_pgsql.rs
use crate::csv::{csv_read, clean_identifier};
use crate::inserter::{ConnParams, Inserter};
use anyhow::{Result, anyhow};
use postgres::{Client, NoTls};
use std::fmt::Write;

pub struct PgSQLInserter {
    client: Client,
}

impl PgSQLInserter {
    pub fn new(p: &ConnParams) -> Result<Self> {
        let dsn = format!(
            "host={} port={} user={} password={} dbname={} sslmode=disable",
            p.host, p.port, p.user, p.password, p.database
        );
        let client = Client::connect(&dsn, NoTls)?;
        Ok(Self { client })
    }

    fn quote(name: &str) -> String {
        let clean = clean_identifier(name);
        format!("\"{}\"", clean.replace('"', "\"\""))
    }

    fn build_cols(headers: &[String]) -> String {
        headers.iter()
            .map(|h| Self::quote(h))
            .collect::<Vec<_>>()
            .join(", ")
    }

    fn build_placeholders(count: usize) -> String {
        (1..=count)
            .map(|i| format!("${}", i))
            .collect::<Vec<_>>()
            .join(", ")
    }

    fn escape_csv_field(val: &str) -> String {
        if val.contains(',') || val.contains('"') || val.contains('\n') {
            format!("\"{}\"", val.replace('"', "\"\""))
        } else {
            val.to_string()
        }
    }
}

impl Inserter for PgSQLInserter {

    // ─── default_insert ───────────────────────────────────────────────────────

    fn default_insert(&mut self, csv_file: &str, table: &str) -> Result<usize> {
        let data  = csv_read(csv_file)?;
        let ncols = data.headers.len();
        let sql   = format!(
            "INSERT INTO {} ({}) VALUES ({})",
            Self::quote(table),
            Self::build_cols(&data.headers),
            Self::build_placeholders(ncols),
        );

        let stmt = self.client.prepare(&sql)?;

        let mut tx = self.client.transaction()?;
        for row in &data.rows {
            let params: Vec<&(dyn postgres::types::ToSql + Sync)> = row.iter()
                .map(|v| v as &(dyn postgres::types::ToSql + Sync))
                .collect();
            tx.execute(&stmt, &params)?;
        }
        tx.commit()?;
        Ok(data.rows.len())
    }

    // ─── bulk_insert ──────────────────────────────────────────────────────────

    fn bulk_insert(&mut self, csv_file: &str, table: &str,
                   batch_size: usize) -> Result<usize> {
        let data   = csv_read(csv_file)?;
        let ncols  = data.headers.len();
        let cols   = Self::build_cols(&data.headers);
        let qtable = Self::quote(table);
        let mut total = 0;

        for chunk in data.rows.chunks(batch_size) {
            let mut sql = format!("INSERT INTO {} ({}) VALUES ", qtable, cols);
            let mut param_idx = 1;

            for (r, _) in chunk.iter().enumerate() {
                if r > 0 { sql.push_str(", "); }
                sql.push('(');
                for c in 0..ncols {
                    if c > 0 { sql.push_str(", "); }
                    write!(sql, "${}", param_idx).unwrap();
                    param_idx += 1;
                }
                sql.push(')');
            }

            let params: Vec<&(dyn postgres::types::ToSql + Sync)> = chunk.iter()
                .flat_map(|row| row.iter()
                    .map(|v| v as &(dyn postgres::types::ToSql + Sync)))
                .collect();

            let mut tx = self.client.transaction()?;
            tx.execute(sql.as_str(), &params)?;
            tx.commit()?;
            total += chunk.len();
        }

        Ok(total)
    }

    // ─── file_insert — COPY FROM STDIN ────────────────────────────────────────

    fn file_insert(&mut self, csv_file: &str, table: &str) -> Result<usize> {
        let data = csv_read(csv_file)?;

        // Формируем CSV в памяти
        let mut buf = String::new();
        let header_line = data.headers.iter()
            .map(|h| Self::escape_csv_field(h))
            .collect::<Vec<_>>()
            .join(",");
        buf.push_str(&header_line);
        buf.push('\n');

        for row in &data.rows {
            let line = row.iter()
                .map(|v| Self::escape_csv_field(v))
                .collect::<Vec<_>>()
                .join(",");
            buf.push_str(&line);
            buf.push('\n');
        }

        let copy_sql = format!(
            "COPY {} FROM STDIN WITH (FORMAT csv, HEADER true)",
            Self::quote(table)
        );

        let mut writer = self.client
            .copy_in(&copy_sql)
            .map_err(|e| anyhow!("COPY error: {}", e))?;

        std::io::Write::write_all(&mut writer, buf.as_bytes())
            .map_err(|e| anyhow!("COPY write error: {}", e))?;

        writer.finish()
            .map_err(|e| anyhow!("COPY finish error: {}", e))?;

        Ok(data.rows.len())
    }
}